File: Project1/project/python/klotski/board.py
import copy

from Vector import vector
from klotski import block


class Board:
    blocks = []
    board = [[' ', '0', '0', ' '],
             [' ', '0', '0', ' '],
             [' ', ' ', ' ', ' '],
             [' ', ' ', ' ', ' '],
             [' ', ' ', ' ', ' ']]

    moveCount = 0

    def __init__(self):
        self.addBlock(0, 0, 1)
        self.blocks[0].width = 2
        self.blocks[0].height = 2

    def addBlock(self, bid, x, y):
        exists = 0
        i = 0
        self.board[x][y] = bid
        while i < len(self.blocks):
            if self.blocks[i].bid is bid:
                if x > self.blocks[i].x:
                    self.blocks[i].makeVertical()
                else:
                    self.blocks[i].makeHorizontal()
                exists = 1
            i += 1

        if exists is 0:
            aux = block.Block(bid, x, y)
            self.blocks.append(aux)

    def goal_test(self):
        if self.blocks[0].getCoords() is [3, 1]:
            return True
        else:
            return False

    def canMove(self, bid, direction):
        bid = int(bid)
        if 0 > bid <= len(self.blocks):
            return False
        block = self.blocks[bid]
        if block is None:
            return False

        if direction is 'w':
            if block.getX() > 0:
                if self.board[int(block.getX()) - 1][int(block.getY())] is ' ':
                    if block.getWidth() is 2:
                        if self.board[int(block.getX()) - 1][int(block.getY()) + 1] is ' ':
                            return True
                        else:
                            return False
                    else:
                        return True
                else:
                    return False
            else:
                return False
        elif direction is 'a':
            if block.getY() > 0:
                if self.board[int(block.getX())][int(block.getY()) - 1] is ' ':
                    if block.getHeight() is 2:
                        if self.board[int(block.getX()) + 1][int(block.getY()) - 1] is ' ':
                            return True
                        else:
                            return False
                    else:
                        return True
                else:
                    return False
            else:
                return False
        elif direction is 's':
            if block.getHeight() is 1:
                if block.getX() < 4:
                    if self.board[int(block.getX()) + 1][int(block.getY())] is ' ':
                        if block.getWidth() is 2:
                            if self.board[int(block.getX()) + 1][int(block.getY()) + 1] is ' ':
                                return True
                            else:
                                return False
                        else:
                            return True
                    else:
                        return False
                else:
                    return False
            else:
                if block.getX() < 3:
                    if self.board[int(block.getX()) + 2][int(block.getY())] is ' ':
                        if block.getWidth() is 2:
                            if self.board[int(block.getX()) + 2][int(block.getY()) + 1] is ' ':
                                return True
                            else:
                                return False
                        else:
                            return True
                    else:
                        return False
                else:
                    return False
        elif direction is 'd':
            if block.getWidth() is 1:
                if block.getY() < 3:
                    if self.board[int(block.getX())][int(block.getY()) + 1] is ' ':
                        if block.getHeight() is 2:
                            if self.board[int(block.getX()) + 1][int(block.getY()) + 1] is ' ':
                                return True
                            else:
                                return False
                        else:
                            return True
                    else:
                        return False
                else:
                    return False
            else:
                if block.getY() < 2:
                    if self.board[int(block.getX())][int(block.getY()) + 2] is ' ':
                        if block.getHeight() is 2:
                            if self.board[int(block.getX()) + 1][int(block.getY()) + 2] is ' ':
                                return True
                            else:
                                return False
                        else:
                            return True
                    else:
                        return False
                else:
                    return False

    def update(self):
        self.board = [[' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' '],
                      [' ', ' ', ' ', ' ']]
        i = 0
        while i < len(self.blocks):
            curr = self.blocks[i]
            #print(f'bid: {curr.bid}, x: {curr.getX()}, y: {curr.getY()}, Height: {curr.getHeight()}, Width: {curr.getWidth()}')
            if curr.getWidth() is 2:
                coords = curr.getCoords()
                self.board[coords[0]][coords[1]] = str(curr.getBid())
                self.board[coords[0]][coords[1] + 1] = str(curr.getBid())
                if curr.getHeight() is 2:
                    self.board[coords[0] + 1][coords[1]] = str(curr.getBid())
                    self.board[coords[0] + 1][coords[1] + 1] = str(curr.getBid())
            elif curr.getHeight() is 2:
                coords = curr.getCoords()
                self.board[coords[0]][coords[1]] = str(curr.getBid())
                self.board[coords[0] + 1][coords[1]] = str(curr.getBid())
            else:
                coords = curr.getCoords()
                self.board[coords[0]][coords[1]] = str(curr.getBid())

            i += 1

    def move(self, piece, direction):
        piece = int(piece)
        if self.canMove(piece, direction) is True:
            if direction is 'w':
                self.blocks[piece].moveUp()
            elif direction is 's':
                self.blocks[piece].moveDown()
            elif direction is 'a':
                self.blocks[piece].moveLeft()
            elif direction is 'd':
                self.blocks[piece].moveRight()
            self.update()
            self.moveCount += 1
            return True
        else:
            return False

    def heuristics(self):
        return abs(3 - self.blocks[0].getX()) + abs(1 - self.blocks[0].getY())

    def __str__(self):
        string = ''
        i = 0

        while i < 5:
            j = 0
            while j < 4:
                string += str(self.board[i][j]) + ' '
                j += 1
            string += '\n'
            i += 1
        return string

    def __eq__(self, other):
        return self.board == other.board

    def getPossibleMoves(self):
        i = 0
        moves = []
        while i < 5:
            j = 0
            while j < 4:
                if self.board[i][j] is ' ':
                    if i > 0 and self.board[i-1][j] is not ' ':
                        if self.canMove(self.board[i-1][j], 's'):
                            moves.append([self.board[i-1][j], 's'])
                    if i < 4 and self.board[i+1][j] is not ' ':
                        if self.canMove(self.board[i+1][j], 'w'):
                            moves.append([self.board[i+1][j], 'w'])
                    if j > 0 and self.board[i][j-1] is not ' ':
                        if self.canMove(self.board[i][j-1], 'd'):
                            moves.append([self.board[i][j-1], 'd'])
                    if j < 3 and self.board[i][j+1] is not ' ':
                        if self.canMove(self.board[i][j+1], 'a'):
                            moves.append([self.board[i][j+1], 'a'])
                j += 1
            i += 1
        #moves.append([13, 'd'])
        return moves

    def copy_blocks(self):
        newone = type(self)
        newone.blocks[0] = self.blocks[0].copy()
        i = 1
        while i < len(self.blocks):
            aux_block = self.blocks[i].copy()
            newone.blocks.append(aux_block)
            i += 1
        newone.board = self.board[:]
        newone.moveCount = copy.copy(self.moveCount)
        return newone

File: Project1/project/c/board.c
//
// Created by linuxlite on 06-04-2019.
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "board.h"

void addBlock(board *self, int nbid, int x, int y) {
    int bid = -1;
    if (nbid > '9'){
        bid = 10 + (nbid - '1' - '0');
    } else {
        bid = nbid - '0';
    }
    bool exists = false;
    unsigned int i = 0;
    self->board[x][y] = bid;
    block *bl = (block *)at(&(self->blocks), bid);
    if(bl != 0) {
        if (bl->bid == bid) {
            if (x > bl->x) {
                makeVertical(bl);
            } else {
                makeHorizontal(bl);
            }
            exists = true;
        }
    }
    if(!exists) {
        block *aux = (block *)malloc(sizeof(block));
        Block(aux, bid, x, y);
        insert(&(self->blocks), aux, bid);
    }
}

bool goal_test(board *self) {
    block *bl = (block *)at(&(self->blocks), 0);
    if (bl->x == 3 && bl->y == 1)
        return true;
    else
        return false;
}

bool canMove(board *self, unsigned int bid, char direction) {
    if(bid < 0 || bid >= self->blocks.size)
        return false;
    block *bl = (block *)at(&(self->blocks), bid);
    if (bl == 0)
        return false;

    if (direction == 'w') {
        if (bl->x > 0) {
            if (self->board[bl->x - 1][bl->y] == ' ') {
                if (bl->width == 2) {
                    if (self->board[bl->x - 1][bl->y + 1] == ' ')
                        return true;
                    else
                        return false;
                } else {
                    return true;
                }
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
    else if(direction == 'a') {
        if (bl->y > 0) {
            if (self->board[bl->x][bl->y - 1] == ' ') {
                if (bl->height == 2) {
                    if (self->board[bl->x + 1][bl->y - 1] == ' ')
                        return true;
                    else
                        return false;
                } else {
                    return true;
                }
            } else {
                return false;
            }
        } else {
            return false;
        }
    } else if (direction == 's'){
        if ((bl->height + bl->x < 5) && self->board[(bl->height + bl->x)][bl->y] == ' ') {
            if (bl->width == 2) {
                if (self->board[(bl->height + bl->x)][bl->y + 1] == ' ')
                    return true;
                else
                    return false;
            } else {
                return true;
            }
        } else {
            return false;
        }
    } else if (direction == 'd') {
        if ((bl->width + bl->y < 4) && self->board[bl->x][bl->width + bl->y] == ' ') {
            if (bl->height == 2) {
                if (self->board[(bl->x + 1)][bl->width + bl->y] == ' ')
                    return true;
                else
                    return false;
            } else {
                return true;
            }
        } else {
            return false;
        }
    }
}

void update(board *self) {
    char aux[5][4] = {{' ', ' ', ' ', ' '},
                      {' ', ' ', ' ', ' '},
                      {' ', ' ', ' ', ' '},
                      {' ', ' ', ' ', ' '},
                      {' ', ' ', ' ', ' '}};
    memcpy(&(self->board), &aux, 5*4* sizeof(char));
    for(unsigned int i=0; i < self->blocks.size; i++) {
        block *curr = (block *) at(&(self->blocks), i);
        if (curr->width == 2) {
            self->board[curr->x][curr->y] = curr->bid;
            self->board[curr->x][curr->y + 1] = curr->bid;
            if (curr->height == 2) {
                self->board[curr->x + 1][curr->y] = curr->bid;
                self->board[curr->x + 1][curr->y + 1] = curr->bid;
            }
        } else if (curr->height == 2) {
            self->board[curr->x][curr->y] = curr->bid;
            self->board[curr->x + 1][curr->y] = curr->bid;
        } else {
            self->board[curr->x][curr->y] = curr->bid;
        }
    }
}

bool move(board *self, unsigned int piece, char direction){
    if (canMove(self, piece, direction) == true){
        block *bl = (block *)at(&(self->blocks), piece);
        if (direction == 'w'){
            moveUp(bl);
        } else if (direction == 's'){
            moveDown(bl);
        } else if (direction == 'a'){
            moveLeft(bl);
        } else if (direction == 'd'){
            moveRight(bl);
        }
        update(self);
        self->moveCount++;
        return true;
    } else
        return false;
}

unsigned int heuristics(board *self){
    block *bl = (block *)at(&(self->blocks), 0);
    return abs(3 - bl->x) + abs(1 - bl->y);
}

void printState(board *self){
    for(unsigned int i = 0; i < 5; i++){
        for(unsigned int j = 0; j < 4; j++){
            char c;
            if (self->board[i][j] == ' ')
                printf("  ");
            else
                printf("%x ", self->board[i][j]);
        }
        endl;
    }
}

bool compStates(board *self, board *other){
    for (unsigned int x = 0; x < 5; x++){
        for (unsigned int y = 0; y < 4; y++){
            if (self->board[x][y] != other->board[x][y])
                return false;
        }
    }
    return true;
}

vector getPossibleMoves(board *self) {
    VECTOR(moves);
    for(unsigned int i = 0; i < 5; i++){
        for(unsigned int j = 0; j < 4; j++){
            if(self->board[i][j] == ' '){
                if(i > 0 && self->board[i - 1][j] != ' '){
                    if(canMove(self, self->board[i - 1][j], 's')) {
                        movement *mov = (movement *)malloc(sizeof(movement));
                        mov->bid = self->board[i - 1][j];
                        mov->direction = 's';
                        add(&moves, mov);
                    }
                }
                if(i < 4 && self->board[i + 1][j] != ' '){
                    if(canMove(self, self->board[i + 1][j], 'w')) {
                        movement *mov = (movement *)malloc(sizeof(movement));
                        mov->bid = self->board[i + 1][j];
                        mov->direction = 'w';
                        add(&moves, mov);
                    }
                }
                if(j > 0 && self->board[i][j - 1] != ' '){
                    if(canMove(self, self->board[i][j - 1], 'd')) {
                        movement *mov = (movement *)malloc(sizeof(movement));
                        mov->bid = self->board[i][j - 1];
                        mov->direction = 'd';
                        add(&moves, mov);
                    }
                }
                if(j < 3 && self->board[i][j + 1] != ' '){
                    if(canMove(self, self->board[i][j + 1], 'a')) {
                        movement *mov = (movement *)malloc(sizeof(movement));
                        mov->bid = self->board[i][j + 1];
                        mov->direction = 'a';
                        add(&moves, mov);
                    }
                }
            }
        }
    }
    return moves;
}

board * duplicate(board *self) {
    board *newone = (board *)malloc(sizeof(board));

    newone->moveCount = self->moveCount;

    //memcpy(&(newone.board), &(self->board), 4*5* sizeof(char));

    for (unsigned int x = 0; x < 5; x++){
        for (unsigned int y = 0; y < 4; y++){
            newone->board[x][y] = self->board[x][y];
        }
    }

    newone->blocks.size = 0;

    for (unsigned int i = 0; i < self->blocks.size; i++) {
        block *bl = (block *)malloc(sizeof(block));
        (*bl) = copy((block *)at(&(self->blocks), i));
        add(&(newone->blocks), bl);
    }
    return newone;
}

board * BOARD(){
    board *ret = (board *)malloc(sizeof(board));
    char arr[5][4] = {{' ', 0, 0, ' '},{' ', 0, 0, ' '},{' ', ' ', ' ', ' '},{' ', ' ', ' ', ' '},{' ', ' ', ' ', ' '}};
    memcpy(&(ret->board), arr, 5*4* sizeof(char));
    ret->moveCount = 0;
    VECTOR(bl);
    block *bc = (block *)malloc(sizeof(block));
    Block(bc, 0, 0, 1);
    makeSquare(bc);
    add(&bl, bc);
    ret->blocks = bl;
    return ret;
}